wnprc_billing/src/org/labkey/wnprc_billing/WNPRC_BillingController.java

```diff
@@ -472,59 +497,138 @@ public void setActivityID(String activityID)
-            String billingAddress = moduleProperties.get(WNPRC_BillingModule.BillingAddress).getEffectiveValue(getContainer());
-            String creditToAccount = moduleProperties.get(WNPRC_BillingModule.CreditToAccount).getEffectiveValue(getContainer());
+    @RequiresPermission(ReadPermission.class)
+    public class PDFExportAction extends ExportAction<InvoicePdfForm>
+    {
+        @Override
+        public void export(InvoicePdfForm invoicePdfForm, HttpServletResponse response, BindException errors) throws Exception
+        {
+            PDFFile pdfFile = getInvoicePDF(invoicePdfForm.getInvoiceNumber(), invoicePdfForm.getName());
+            PageFlowUtil.prepareResponseForFile(getViewContext().getResponse(), Collections.emptyMap(), pdfFile.getFileName(), invoicePdfForm.isAsAttachment());
+            pdfFile.getInvoicePDF().output(getViewContext().getResponse().getOutputStream());
+        }
+    }
 
-            double tierRate = accountTierRate != null ? accountTierRate.getTierRate() : 0;
-            String formName = invoicePdfForm.getName();
-            InvoicePDF pdf = formName.equals("Invoice PDF") ? new InvoicePDF(invoice, alias, invoiceRun, tierRate, contactEmail, billingAddress, creditToAccount) :
-                    new SummaryPDF(invoice, alias, invoiceRun, tierRate, contactEmail, billingAddress, creditToAccount);
+    private class PDFFile
+    {
+        String _fileName;
+        InvoicePDF _invoicePDF;
 
-            pdf.addPage();
-            List<InvoicedItem> invoicedItems;
+        public String getFileName()
+        {
+            return _fileName;
+        }
 
-            if (formName.equalsIgnoreCase("Invoice PDF"))
-            {
-                invoicedItems = getInvoicedItems(invoicePdfForm.getInvoiceNumber());
-                pdf.createLineItems(invoicedItems, true);
-            }
-            else
-            {
-                invoicedItems = getSummarizedItems(invoicePdfForm.getInvoiceNumber());
-                pdf.createLineItems(invoicedItems, false);
-            }
+        public void setFileName(String fileName)
+        {
+            _fileName = fileName;
+        }
 
-            SimpleDateFormat dateFormatBillingFor = new SimpleDateFormat("MM_yyyy");
-            String filename = alias.getGrantNumber() + "_" + dateFormatBillingFor.format(invoiceRun.getBillingPeriodStart()) + "_Invoice.pdf";
-            PageFlowUtil.prepareResponseForFile(getViewContext().getResponse(), Collections.emptyMap(), filename, invoicePdfForm.isAsAttachment());
-            pdf.output(getViewContext().getResponse().getOutputStream());
+        public InvoicePDF getInvoicePDF()
+        {
+            return _invoicePDF;
+        }
+
+        public void setInvoicePDF(InvoicePDF invoicePDF)
+        {
+            _invoicePDF = invoicePDF;
+        }
+    }
+
+    private PDFFile getInvoicePDF(String invoiceNumber, String formName) throws IOException
+    {
+        Invoice invoice = getInvoice(invoiceNumber);
```

Comment:
This is a security concern. There's no filtering for a matching container. Thus, if you have read access to any container and know the invoice number, you can get its PDF

Reply:
Filter on Container is now added, see getInvoice().